rust/agama-network/src/nm/model.rs

```diff
@@ -0,0 +1,161 @@
+
+/// NetworkManager wireless mode
+///
+/// Using the newtype pattern around an String is enough. For proper support, we might replace this
+/// struct with an enum.
+use crate::{
+    model::{DeviceType, IpMethod, SecurityProtocol, WirelessMode},
+    nm::error::NmError,
+};
+use std::fmt;
+
+#[derive(Debug, PartialEq)]
+pub struct NmWirelessMode(pub String);
+
+impl Default for NmWirelessMode {
+    fn default() -> Self {
+        NmWirelessMode("infrastructure".to_string())
+    }
+}
+
+impl From<&str> for NmWirelessMode {
+    fn from(value: &str) -> Self {
+        Self(value.to_string())
+    }
+}
+
+impl NmWirelessMode {
+    pub fn as_str(&self) -> &str {
+        &self.0.as_str()
+    }
+}
+
+impl From<NmWirelessMode> for WirelessMode {
+    fn from(value: NmWirelessMode) -> Self {
+        match value.as_str() {
+            "infrastructure" => WirelessMode::Infra,
+            "adhoc" => WirelessMode::AdHoc,
+            "mesh" => WirelessMode::Mesh,
+            "ap" => WirelessMode::AP,
+            _ => WirelessMode::Other,
+        }
+    }
+}
+
+/// Device types
+///
+/// As we are using the number just to filter wireless devices, using the newtype
+/// pattern around an u32 is enough. For proper support, we might replace this
+/// struct with an enum.
+#[derive(Debug, Clone, Copy)]
+pub struct NmDeviceType(pub u32);
+
+impl From<NmDeviceType> for u32 {
+    fn from(value: NmDeviceType) -> u32 {
+        value.0
+    }
+}
+
+impl Default for NmDeviceType {
+    fn default() -> Self {
+        NmDeviceType(0)
+    }
+}
+
+impl TryFrom<NmDeviceType> for DeviceType {
+    type Error = NmError;
+
+    fn try_from(value: NmDeviceType) -> Result<Self, Self::Error> {
+        match value {
+            NmDeviceType(0) => Ok(DeviceType::Loopback),
+            NmDeviceType(1) => Ok(DeviceType::Ethernet),
+            NmDeviceType(2) => Ok(DeviceType::Wireless),
+            NmDeviceType(_) => Err(NmError::UnsupportedDeviceType(value.into())),
+        }
+    }
+}
+
+/// Key management
+///
+/// Using the newtype pattern around an String is enough. For proper support, we might replace this
+/// struct with an enum.
+#[derive(Debug, PartialEq)]
+pub struct NmKeyManagement(pub String);
+
+impl Default for NmKeyManagement {
+    fn default() -> Self {
+        NmKeyManagement("none".to_string())
+    }
+}
+
+impl From<&str> for NmKeyManagement {
+    fn from(value: &str) -> Self {
+        Self(value.to_string())
+    }
+}
+
+impl From<NmKeyManagement> for SecurityProtocol {
+    fn from(value: NmKeyManagement) -> Self {
+        match value.as_str() {
+            "owe" => SecurityProtocol::OWE,
+            "ieee8021x" => SecurityProtocol::DynamicWEP,
+            "wpa-psk" => SecurityProtocol::WPA2,
+            "wpa-eap" => SecurityProtocol::WPA3Personal,
+            "sae" => SecurityProtocol::WPA2Enterprise,
+            "wpa-eap-suite-b192" => SecurityProtocol::WPA2Enterprise,
+            _ => SecurityProtocol::WEP,
```

Comment:
so wep does not have own string? It kind of smells for me. I expect that "wep" has something and report error for something we do not know.